src/Workspaces/Core/Portable/Options/OptionServiceFactory.cs

```diff
@@ -125,20 +117,28 @@ public OptionSet GetOptions()
 
             public void RegisterDocumentOptionsProvider(IDocumentOptionsProvider documentOptionsProvider)
             {
-                lock (_gate)
+                InterlockedUpdate(
+                    ref _documentOptionsProviders,
+                    (original, added) => original.Add(added),
+                    documentOptionsProvider);
+            }
+
+            private static void InterlockedUpdate<T, TArg>(ref ImmutableArray<T> array, Func<ImmutableArray<T>, TArg, ImmutableArray<T>> update, TArg argument)
```

Comment:
❓ I was surprised to find this not already a part of `ImmutableInterlocked`. Do we have a good place to put this method in the current code base?